File: Sources/TileHashC/tile_hash.c
#include "tile_hash.h"
#include <string.h>

TileHashResult tile_hash(const void *data,
                         int startX, int startY,
                         int endX,   int endY,
                         int bytesPerRow) {
    uint64_t hashLo = UINT64_C(5163949831757626579);
    uint64_t hashHi = UINT64_C(4657936482115123397);
    const uint64_t primeLo = UINT64_C(1238197591667094937);
    const uint64_t primeHi = UINT64_C(1700294137212722571);

    const int pixelCount = endX - startX;
    const int pairCount  = pixelCount >> 1;
    const int trailing   = pixelCount & 1;

    const unsigned char *base = (const unsigned char *)data;

    for (int y = startY; y < endY; y++) {
        const unsigned char *p = base + (size_t)y * bytesPerRow + (size_t)startX * 4;

        for (int i = 0; i < pairCount; i++) {
            uint64_t v;
            memcpy(&v, p, 8);
            hashLo = (hashLo ^ v) * primeLo;
            hashHi = (hashHi ^ v) * primeHi;
            p += 8;
        }

        if (trailing) {
            uint32_t v32;
            memcpy(&v32, p, 4);
            uint64_t v = (uint64_t)v32;
            hashLo = (hashLo ^ v) * primeLo;
            hashHi = (hashHi ^ v) * primeHi;
        }
    }

    TileHashResult result;
    result.hashLo = (int64_t)hashLo;
    result.hashHi = (int64_t)hashHi;
    return result;
}

File: Sources/TileHashSwift/TileHash.swift
public struct TileHashResult {
    public let hashLo: Int64
    public let hashHi: Int64
}

public func tileHash(ptr: UnsafePointer<UInt8>, startX: Int, startY: Int, endX: Int, endY: Int, bytesPerRow: Int) -> TileHashResult {
    var hashLo: Int64 = 5_163_949_831_757_626_579
    var hashHi: Int64 = 4_657_936_482_115_123_397
    let primeLo: Int64 = 1_238_197_591_667_094_937
    let primeHi: Int64 = 1_700_294_137_212_722_571

    let raw = UnsafeRawPointer(ptr)
    let pixelCount = endX &- startX
    let pairCount = pixelCount >> 1
    let hasTrailingPixel = pixelCount & 1 != 0

    for y in startY..<endY {
        var p = raw + (y &* bytesPerRow &+ startX &* 4)

        for _ in 0..<pairCount {
            let v = Int64(bitPattern: p.loadUnaligned(as: UInt64.self))
            hashLo = (hashLo ^ v) &* primeLo
            hashHi = (hashHi ^ v) &* primeHi
            p += 8
        }

        if hasTrailingPixel {
            let v = Int64(p.loadUnaligned(as: UInt32.self))
            hashLo = (hashLo ^ v) &* primeLo
            hashHi = (hashHi ^ v) &* primeHi
        }
    }

    return TileHashResult(hashLo: hashLo, hashHi: hashHi)
}
